ToolBox/SmartInsider/SmartInsiderConverter.cs

```diff
@@ -0,0 +1,288 @@
+
+                    // Log any mapping events since this can be a point of failure
+                    if (ticker != newTicker)
+                    {
+                        Log.Trace($"SmartInsiderConverter.Process(): Mapped ticker from {ticker} to {newTicker}");
+                    }
+
+                    List<T> symbolLines;
+                    if (!lines.TryGetValue(newTicker, out symbolLines))
+                    {
+                        symbolLines = new List<T>();
+                        lines[newTicker] = symbolLines;
+                    }
+
+                    symbolLines.Add(dataInstance);
+
+                    previousMarket = dataInstance.TickerCountry;
+                    previousTicker = ticker;
+                }
+                catch (Exception e)
+                {
+                    Log.Error(e, $"SmartInsiderConverter.Process(): Error on line {i}");
+                }
+            }
+
+            return lines;
+        }
+
+        /// <summary>
+        /// Writes to a temp file and moves the content to the final directory
+        /// </summary>
+        /// <param name="finalFile">Final file to write to</param>
+        /// <param name="contents">Contents to write to file</param>
+        private void WriteToFile<T>(DirectoryInfo destinationDirectory, Dictionary<string, List<T>> contents)
+            where T : SmartInsiderEvent
+        {
+            foreach (var kvp in contents)
+            {
+                var ticker = kvp.Key.ToLower();
+
+                var finalFile = new FileInfo(Path.Combine(destinationDirectory.FullName, $"{ticker}.csv"));
+                var fileContents = new List<T>();
+
+                if (finalFile.Exists)
+                {
+                    Log.Trace($"SmartInsiderConverter.WriteToFile(): Writing to existing file: {finalFile.FullName}");
+                    fileContents = File.ReadAllLines(finalFile.FullName)
+                        .Select(x => (T)CreateSmartInsiderInstance<T>(x))
+                        .ToList();
+                }
+                else
+                {
+                    Log.Trace($"SmartInsiderConverter.WriteToFile(): Writing to new file: {finalFile.FullName}");
+                }
+
+                fileContents.AddRange(kvp.Value);
+
+                var csvContents = fileContents
+                    .OrderBy(x => x.LastUpdate)
+                    .Select(x => x.ToCsv())
+                    .Distinct();
+
+                File.WriteAllLines(finalFile.FullName, csvContents);
+            }
+        }
+
+        /// <summary>
+        /// Resolves type parameter to corresponding <see cref="SmartInsiderEvent"/> derived class
+        /// </summary>
+        /// <typeparam name="T"><see cref="SmartInsiderEvent"/> derived class</typeparam>
+        /// <param name="line">CSV line</param>
+        /// <returns>SmartInsiderEvent derived class</returns>
+        private SmartInsiderEvent CreateSmartInsiderInstance<T>(string line)
```

Comment:
I believe you can add `: where T : SmartInsiderEvent` and make the return type `T` -- this will provide full static typing at the call site 👍